### PR TITLE
Work around Closure compiler issue with unicode characters

### PR DESCRIPTION
**Pull request by juj**
Work around Closure compiler issue https://github.com/google/closure-compiler/issues/3784 that prevents calling Closure compiler with a cmdline that contains unicode characters.

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -7468,6 +7468,14 @@ def test_closure_full_js_library(self):
   def test_closure_externs(self):
     self.run_process([EMCC, test_file('hello_world.c'), '--closure=1', '--pre-js', test_file('test_closure_externs_pre_js.js'), '--closure-args', '--externs "' + test_file('test_closure_externs.js') + '"'])
 
+  # Tests that it is possible to enable the Closure compiler via --closure=1 even if any of the input files reside in a path with unicode characters.
+  def test_closure_cmdline_utf8_chars(self):
+    test = "☃ äö Ć € ' 🦠.c"
+    shutil.copyfile(test_file('hello_world.c'), test)
+    externs = '💩' + test
+    open(externs, 'w').write('')
```
Comment (sbc100) [inline tests/test_other.py:7476]: You can replace these two lines with `create_file` I think.

File: tools/building.py [hunk 2]
```diff
@@ -1006,19 +1004,43 @@ def add_to_path(dirname):
...
+    # Closure compiler does not work if any of the input files contain characters outside the
+    # 7-bit ASCII range. Therefore make sure the command line we pass does not contain any such
+    # input files by passing all input filenames relative to the cwd. (user temp directory might
+    # be in user's home directory, and user's profile name might contain unicode characters)
+    cwd = os.getcwd()
+    try:
+      os.chdir(configuration.get_temp_files_directory())
+      proc = run_process(cmd, stderr=PIPE, check=False, env=env)
+    finally:
+      os.chdir(cwd)
```
Comment (sbc100) [inline tools/building.py:1044]: You can do `with utils.chdir():` instead of the try/catch.

File: tools/building.py [hunk 3]
```diff
@@ -1006,19 +1004,43 @@ def add_to_path(dirname):
     # Tell closure never to inject the 'use strict' directive.
     args += ['--emit_use_strict=false']
 
+    # Closure compiler is unable to deal with path names that are not 7-bit ASCII:
+    # https://github.com/google/closure-compiler/issues/3784
+    outfile = configuration.get_temp_files().get('.cc.js').name  # Safe 7-bit filename
```
Comment (sbc100) [inline tools/building.py:1009]: I think its best to minimize the number of calls to `get_temp_files()` and share the return value instead.  

```
tmpfiles = configuration.get_temp_files()
outfile = tempfiles.get('.cc.js').name
```

As well as doing less work this also prints less stuff to the log (each TempFiles instance prints `not cleaning up temp files since in debug-save mode, see them ..` to stdout in save temps mode).   We should really fix that but that is completely unrelated to this change.

You can also then do `tempfiles.tmpdir` below to avoid changing `shared.py` at all.

File: tools/building.py (hunk 2 above)
Comment (sbc100) [reply tools/building.py:1044]: Actuall you don't even need that, you can just pass `cwd=` to run_process.

Comment (juj) [reply tools/building.py:1044]: Oh right, good catch!

File: emcc.py [hunk 4]
```diff
@@ -2902,9 +2902,11 @@ def run_closure_compiler():
                                use_closure_compiler=options.use_closure_compiler,
                                debug_info=debug_info,
                                symbols_file=symbols_file)
+
+    shared.configuration.get_temp_files().note(wasm2js)
+
```
Comment (sbc100) [inline emcc.py:2907]: This seems wrong, isn't this file used down on line 2912?   (maybe an accident?  or unrelated change?)

**Review by sbc100 — APPROVED** [review]
nice!

File: emcc.py (hunk 4 above)
Comment (juj) [reply emcc.py:2907]: This line only causes the temp file to be marked for deletion when program exits. It is fine to use it still (we do have that pattern in other places as well)

Although I am not entirely sure why `wasm2js.js` and `wasm2js.modular.js` all of a sudden are not getting deleted, but looking at these code sites, they are leaving behind the temp files they generate.

Comment (sbc100) [reply emcc.py:2907]: But we run the test suite in such way that if any files are left around the test fail...  

At the very least this is unrelated to fix in question isn't it?

Comment (juj) [reply emcc.py:2907]: Yeah, that is exactly what is happening. After the refactor of Closure, the test suite began failing on these temp files getting leaked, and I needed to add these in. I don't quite understand why, but these `.note()`s are certainly the proper thing to do as they are missing from the calls to wasm2js and the modular version of it.

Comment (sbc100) [reply emcc.py:2907]: I see.  That that is very odd.   

Isn't there a case where the wasm2js output here actually *is* the final output?  The line `final_js = wasm2js` below suggests it could be.  

At least we should understand why refactoring the closure stuff caused this issue.   was the the case the closure cleaned this up before?   Or that closure was operating in-place before and now its not?

Comment (juj) [reply emcc.py:2907]: >  Or that closure was operating in-place before and now its not?

I believe that is the case here.

> Isn't there a case where the wasm2js output here actually _is_ the final output? The line `final_js = wasm2js` below suggests it could be.

I'm letting the CI tell me if it's so, I believe it should bark on something in a couple of minutes.

**Comment by juj** [comment]
Nope, looks like all good.
